### PR TITLE
Ruby group capstone - Catalog of my things

### PR DESCRIPTION
**Pull request by brighton0102**
### In this milestone I : 

[ ] Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) should be created in a separate .rb file.
[ ] All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
[ ] All properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
[ ] All methods visible in the diagram should be implemented:

1. add_item method (in all classes that have that method)
2. should take an instance of the Item class as an input.
3. should add the input item to the collection of items.
4. should add self as a property of the item object (by using the correct setter from the item object).

- [ ] can_be_archived?() in the Item class

1. should return true if published_date is older than 10 years.
2. otherwise, it should return false.

- [ ] move_to_archive() in the Item class

1.  should reuse can_be_archived?() method.
2.  should change the archived property to true if the result of the can_be_archived?() method is true.
3.  should do nothing if the result of the can_be_archived?() method is false.

- [ ] can_be_archived?() in the Book class

1. [ ] should override the method from the parent class.
2. [ ] should return true if parent's method returns true OR if cover_state equals to "bad".
3. [ ] otherwise, it should return false.

- [ ] can_be_archived?() in the MusicAlbum class

1. should override the method from the parent class.
2. should return true if parent's method returns true AND if on_spotify equals true.
3. otherwise, it should return false.

- [ ] can_be_archived?() in the Movie class

1. should override the method from the parent class.
2. should return true if parent's method returns true OR if silent equals true.
3. otherwise, it should return false.

- [ ] can_be_archived?() in the Game class

1. should override the method from the parent class.
2. should return true if parent's method returns true AND if last_played_at is older than 2 years.
3. otherwise, it should return false.

- [ ] Add unit tests for all implemented methods.

image

### CODE REVIEW

File: classes/genre.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+  def initialize(id, name)
+    @id = id
+    @name = name
+    @items = []
+  end
+
+  def add_item(item)
+    @items << item
+  end
```
Comment (nahnahdev) [inline classes/genre.rb:15]: - [ ] Great job so far @brighton0102 There are just a few changes I will suggest you implement and you'll be on your way to the next Milestone :wink: 
- [ ] I noticed here that your relationship between `Genre` and `Items` is not defined both ways. To fix this you need to adapt your solution to manage both sides of the relationship in both cases. You need to modify your implementations of `add_item` in `Genre` and create a setter for `@genre` in `Item`.
- Example
```ruby
    def add_item(item)
      @items.push(item)
      item.genre = self
    end
```

File: classes/author.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  def initialize(id, first_name, last_name)
+    @id = id
+    @first_name = first_name
+    @last_name = last_name
+    @items = []
+  end
+
+  def add_item(item)
+    @items << item
+  end
```
Comment (nahnahdev) [inline classes/author.rb]: - [ ] Just as I mentioned in [this comment](https://github.com/brighton0102/Catalog-of-my-things/pull/16#discussion_r1451576533), I will suggest you implement similar changes here as well :+1: 
- [ ] The relationship between `Author` and `Items` is not defined both ways. To fix this you need to adapt your solution to manage both sides of the relationship in both cases. You need to modify your implementations of `add_item` in `Author` and create a setter for `@author` in `Item`.
- Example
```ruby
    def add_item(item)
      @items.push(item)
      item.author = self
    end
```

File: classes/label.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  def initialize(id, title, color)
+    @id = id
+    @title = title
+    @color = color
+    @items = []
+  end
+
+  def add_item(item)
+    @items << item
+  end
```
Comment (nahnahdev) [inline classes/label.rb:16]: - [ ] A similar change also needs to be implemented here. You can see that your relationship between `Label` and `Items` is not defined both ways. To fix this you need to adapt your solution to manage both sides of the relationship in both cases. You need to modify your implementations of `add_item` in `Label` and create a setter for `@label` in `Item`.
- Example
```ruby
    def add_item(item)
      @items.push(item)
      item.label = self
    end
```

File: classes/item.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# # Item class represents a generic catalog item.
+
+class Item
+  attr_accessor :title, :publish_date
+  attr_reader :id, :archived
```
Comment (nahnahdev) [inline classes/item.rb]: - [ ] Kindly ensure that your `Item Class` includes `attr_reader` for these attributes. :wink: 

```ruby
  attr_reader :genre
  attr_reader :author
  attr_reader :label
```

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+puts 'Welcome to the Catalogue of my things'
+
+def print_options
+  puts "\nOptions"
+  puts '1. List all books'
+  puts '2. List all music albums'
+  puts '3. List all games'
+  puts '4. List all genres'
+  puts '5. List all labels'
+  puts '6. List all authors'
```
Comment (nahnahdev) [inline main.rb:27]: - [ ] Great job @brighton0102 Almost all your console commands work as expected. I only encountered an error when I tried to view all `Authors` by accessing option `6` but this threw an error and shut down the app. So kindly make those changes so your app can display all authors and work as expected :wink: 
- [ ] Error Message
image

File: operations/file_operations.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    'Label' => Label,
+    'MusicAlbum' => MusicAlbum,
+    'Genre' => Genre,
+    'Game' => Game,
+    'Author' => Author
+  }
+  json_data = JSON.parse(File.read('items.json'))
+  json_data.map do |item_data|
+    class_name = item_data['class_name']
+    class_mapping[class_name].from_json(item_data)
```
Comment (nahnahdev) [inline operations/file_operations.rb]: - [ ] I also noticed that once `Quit` the console app and I tried to restart the app to check if Data was preserved and to also check whether your method to get back all the data that was created indeed works as expected, the app didn't start at all. So kindly fix that as well :+1: I have attached the error message below.
- [ ] So kindly ensure that Data is well Preserved in your console app and stored data from the `.json files` is loaded when the app starts. Kindly ensure that your app also starts without any errors :wink: 
- Error Message
image

**Review by nahnahdev — CHANGES_REQUESTED** [review]
## Project Status: CHANGES REQUIRED :recycle: 
Hi @brighton0102 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights :fire: 
- Linters well setup and passing :heavy_check_mark: 
- Presentation video included and well presented :heavy_check_mark: 
- Most project requirements implemented :heavy_check_mark: 
- All written tests passing :heavy_check_mark: 
- SQL tables present in the project :heavy_check_mark: 
- Console app works almost completely :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
- N/A

Cheers ,and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  puts '7. Add a book'
+  puts '8. Add a music album'
+  puts '9. Add a game'
+  puts '10. Quit'
+end
```
Comment (Gambit142) [inline main.rb:32]: - [ ] Great work making most of the required changes, however, I noticed that when I try to _**Add a book**_ or try to _**Add a game**_, I get errors in the console application that you created. Kindly fix these errors. 👍 😎 

#### Error When trying to add a book

image

#### Error when trying to add a game

image

#### N.B: Please ensure that all options of the console application work properly after fixing the errors and that a user can save data to JSON files and load data fro the JSON files upon reentry. 💻 🖱️

**Review by Gambit142 — CHANGES_REQUESTED** [review]
Hi @brighton0102 👋, 

Great work on making the changes requested by a previous reviewer 👏 

## Highlights
- All linters checks passed ✔️ 
- Created classes ✔️ 
- Created SQL tables using schema diagram ✔️ 

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions

N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer and addressing them will improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Reem-lab — APPROVED** [review]
# Hi Team,  👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights

- [ ]  Descriptive PR⭐
- [ ]  No Linters error 🤩
- [ ] Good job working on the saving data into JSONs💪🏻
- [ ] Amazing job working on the main file 🔥🔥
- [ ] Great for making the requested changes ⭐
- [ ] All options working fine 🔥

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**